### PR TITLE
Tests for population generation

### PR DESCRIPTION
**Pull request by lynn**
This PR adds three test cases:

* `TEST(PopulationGeneration, GeneratedPopulationFitsModel)` generates a population and asserts that it fits the model it was generated from.

* `TEST(PopulationGeneration, GeneratedPopulationIsInfectious)` is an integration tests: it runs a brief simulation with a generated population and asserts that the simulated disease acutally spreads through this population.

* `TEST(GeoPosition, SampleCalculation)` checks a calculation of the distance between two GeoPositions. (There was a bug in the `GeoPosition` distance calculation used by the generator, so I’ve replaced it with some code that uses `boost::geometry` instead of reinventing the wheel.)

I added some more logging calls to the population generator, and split some code from `PopulationBuilder.cpp` into a new method called `population::Generator::FromConfig`.

### CODE REVIEW

File: src/main/cpp/util/InstallDirs.cpp [hunk 1]
```diff
@@ -170,19 +171,36 @@ path InstallDirs::GetRootDir()
...
 	const boost::filesystem::path& anchor_path,
 	boost::property_tree::ptree& result)
 {
-	auto file_path = InstallDirs::GetDataDir() /= relative_path;
-	if (!is_regular_file(file_path)) {
-		FATAL_ERROR(
-			"File " + relative_path.string() +
-			" not present. (full path: " + file_path.string() + ")");
-	}
+	boost::property_tree::read_xml(*OpenFile(relative_path, anchor_path), result);
```
Comment (jonathanvdc) [inline src/main/cpp/util/InstallDirs.cpp]: `OpenFile` returns an `std::unique_ptr<boost::filesystem::ifstream>` right? I'm not sure if it's safe to dereference an `std::unique_ptr<boost::filesystem::ifstream>`. I have a hunch that `OpenFile(...)`'s lifetime ends when the expression it is used in (i.e., the pointer dereference, *not* the call to `read_xml`) ends. If I'm right, then `read_xml` is accessing a pointer to memory that's already been deallocated, which is UB.

File: src/main/cpp/util/InstallDirs.cpp [hunk 2]
```diff
@@ -170,19 +171,36 @@ path InstallDirs::GetRootDir()
 	return g_root_dir;
 }
 
+std::unique_ptr<boost::filesystem::ifstream> InstallDirs::OpenFile(
```
Comment (jonathanvdc) [inline src/main/cpp/util/InstallDirs.cpp:174]: Nice refactor!

File: src/main/cpp/util/InstallDirs.cpp (hunk 1 above)
Comment (jonathanvdc) [reply src/main/cpp/util/InstallDirs.cpp]: Also, thanks for fixing my bug! :)

File: src/test/cpp/gtester/GeoPosition.cpp [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+#include <geo/GeoPosition.h>
+
+namespace Tests {
+
+TEST(GeoPosition, SampleCalculation)
+{
+    using namespace stride::geo;
+    const GeoPosition amsterdam{52.37, 4.90};
+    const GeoPosition paris{48.86, 2.35};
+	ASSERT_NEAR(amsterdam.Distance(paris), 429.7, 1.0);
```
Comment (jonathanvdc) [inline src/test/cpp/gtester/GeoPosition.cpp]: The `ASSERT_NEAR`'s indentation doesn't seem quite right. Has this file been `clang-format`ted?

File: src/test/cpp/gtester/PopulationGeneration.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+TEST(PopulationGeneration, GeneratedPopulationIsInfectious)
+{
+	auto log = spdlog::stderr_logger_st("test_popgen");
+	log->set_level(spdlog::level::off);
+	auto sim = stride::SimulatorBuilder::Build("../config/run_test_popgen.xml", log, 1, false);
+
+    // Run the simulation for 10 days, and assert an increase in infected persons.
+    unsigned int before = sim->GetPopulation()->get_infected_count();
+	for (int i = 0; i < 10; i++)
```
Comment (jonathanvdc) [inline src/test/cpp/gtester/PopulationGeneration.cpp]: The indentation seems a little off here.

File: src/main/cpp/pop/Generator.h [hunk 5]
```diff
@@ -28,52 +30,74 @@ class Generator
...
 	Population Generate();
 
 	/// Check if a population fits the model.
 	/// If verbose is true, log the checks performed.
-	bool FitsModel(const Population& population, bool verbose = false);
+	bool FitsModel(const Population& population);
+
+	/// Log a debug message.
+	template <typename... Args>
+	void Debug(const char* fmt, const Args&... args)
```
Comment (jonathanvdc) [inline src/main/cpp/pop/Generator.h:54]: Nice refactor!

File: src/main/resources/config/run_test_popgen.xml [hunk 6]
```diff
@@ -1,13 +1,15 @@
...
     <r0>11</r0>
     <seeding_rate>0.002</seeding_rate>
     <immunity_rate>0.8</immunity_rate>
-    <geodistribution_profile>belgium_population.csv</geodistribution_profile>
-    <population_file>population_model_default.xml</population_file>
+    <geodistribution_profile>belgium_population_major.csv</geodistribution_profile>
+    <population_file>population_model_test.xml</population_file>
     <reference_households>households_flanders.json</reference_households>
-    <num_days>50</num_days>
+    <num_days>10</num_days>
```
Comment (jonathanvdc) [inline src/main/resources/config/run_test_popgen.xml:12]: Why'd you change the number of days? Are we cooking the books here? :wink:

**Review by jonathanvdc — APPROVED** [review]
Looking good!

File: src/main/cpp/util/InstallDirs.cpp (hunk 1 above)
Comment (lynn) [reply src/main/cpp/util/InstallDirs.cpp]: Yeah, I wasn’t so sure about this either. I’ll extend `OpenFile(…)`’s lifetime.

File: src/main/cpp/util/InstallDirs.cpp [hunk 7]
```diff
@@ -170,19 +171,37 @@ path InstallDirs::GetRootDir()
...
 	boost::property_tree::ptree& result)
 {
-	auto file_path = InstallDirs::GetDataDir() /= relative_path;
-	if (!is_regular_file(file_path)) {
-		FATAL_ERROR(
-			"File " + relative_path.string() +
-			" not present. (full path: " + file_path.string() + ")");
-	}
+	auto stream = OpenFile(relative_path, anchor_path);
+	boost::property_tree::read_xml(*stream, result);
```
Comment (lynn) [inline src/main/cpp/util/InstallDirs.cpp:198]: This should be OK, I think.

File: src/main/resources/config/run_test_popgen.xml (hunk 6 above)
Comment (lynn) [reply src/main/resources/config/run_test_popgen.xml:12]: Haha~ I figured a *population-generation-testing* config file shouldn’t spend so long running a simulation! Verifying that the simulation runs at all is fine.

File: src/main/cpp/util/InstallDirs.cpp (hunk 7 above)
Comment (jonathanvdc) [reply src/main/cpp/util/InstallDirs.cpp:198]: LGTM

**Pull request merged**